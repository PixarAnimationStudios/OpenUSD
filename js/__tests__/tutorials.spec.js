const Module = require("../bindings/Release/jsBindings");
const fs = require("fs");
const path = require("path");

let Usd;

describe('USD Tutorials', () => {
  beforeEach(async () => {
      Usd = await Module();
    }, 500000);

  afterEach(() => {
    Usd.PThread.runningWorkers.forEach(x => x.onmessage = function() {});
    Usd.PThread.terminateAllThreads();
    Usd = null;
    process.removeAllListeners('unhandledRejection')
    process.removeAllListeners('uncaughtException')
  });

  it("Referencing Layers", () => {
    const usdFilename = 'HelloWorld.usda';
    const stage = Usd.UsdStage.CreateNew(usdFilename);
    const xform = stage.DefinePrim('/hello', 'Xform');
    const sphere = stage.DefinePrim('/hello/world', 'Sphere');

    stage.SetDefaultPrim(xform);
    const xformable = new Usd.UsdGeomXformable(xform);
    const translateOp = xformable.AddTranslateOp();
    translateOp.Set([4, 5, 6]);

    const radiusAttr = sphere.GetAttribute('radius');
    radiusAttr.Set(2);

    const color = sphere.GetAttribute("primvars:displayColor");
    color.Set([[0,0,1]]);

    // First Reference
    const refStage = Usd.UsdStage.CreateNew('RefExample.usda')

    refSphere = refStage.OverridePrim('/refSphere')
    refSphere.GetReferences().AddReference('./HelloWorld.usda')
    refXform = new Usd.UsdGeomXformable(refSphere);
    refXform.SetXformOpOrder([]);

    // Second Reference
    const refSphere2 = refStage.OverridePrim('/refSphere2')
    refSphere2.GetReferences().AddReference('./HelloWorld.usda')
    const overSphere = refStage.GetPrimAtPath('/refSphere2/world');
    overSphere.GetAttribute("primvars:displayColor").Set( [[1, 0, 0]] );

    expect(refStage.ExportToString()).toMatchSnapshot();
    expect(refStage.GetRootLayer().ExportToString()).toMatchSnapshot();
  });

  it("Simple Shading in USD", () => {
    const fileName = "SimpleShading.usda";

    const data = fs.readFileSync(path.resolve(__dirname, './test-data/simpleshading.usda'));

    Usd.FS.createDataFile('/', fileName, data, true, true, true);
    const stage = Usd.UsdStage.Open(fileName);

    const stReader = Usd.UsdShadeShader.Define(stage, '/TexModel/boardMat/stReader');
    stReader.CreateIdAttr('UsdPrimvarReader_float2');
    stReader.CreateInput('varname', Usd.ValueTypeNames.Token).ConnectToSourcePath("/TexModel/boardMat.inputs:frame:stPrimvarName");

    const sampler = Usd.UsdShadeShader.Define(stage, "/TexModel/boardMat/texture");
    sampler.CreateIdAttr("UsdUVTexture");
    const fileInput = sampler.CreateInput("file", Usd.ValueTypeNames.Asset);
    fileInput.Set("USDLogoHeader.png");
    const st = sampler.CreateInput("st", Usd.ValueTypeNames.Float2);
    st.ConnectToSource(stReader, 'result')
    sampler.CreateOutput("rgb", Usd.ValueTypeNames.Float3);

    const pbrShader = Usd.UsdShadeShader.Define(stage, '/TexModel/boardMat/PBRShader')
    pbrShader.CreateIdAttr("UsdPreviewSurface");
    pbrShader.CreateInput("roughness", Usd.ValueTypeNames.Float).Set(0.4);
    pbrShader.CreateInput("metallic", Usd.ValueTypeNames.Float).Set(0.0);
    pbrShader.CreateInput("diffuseColor", Usd.ValueTypeNames.Color3f).ConnectToSource(sampler, "rgb");
    pbrShader.CreateOutput("surface", Usd.ValueTypeNames.Token);

    expect(stage.ExportToString()).toMatchSnapshot();
  });
});
