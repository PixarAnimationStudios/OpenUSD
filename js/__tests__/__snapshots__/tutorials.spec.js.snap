// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`USD Tutorials Referencing Layers 1`] = `
"#usda 1.0
(
    doc = \\"\\"\\"Generated from Composed Stage of root layer //RefExample.usda
\\"\\"\\"
)

def Xform \\"refSphere\\"
{
    double3 xformOp:translate = (4, 5, 6)
    uniform token[] xformOpOrder = []

    def Sphere \\"world\\"
    {
        color3f[] primvars:displayColor = [(0, 0, 1)]
        double radius = 2
    }
}

def Xform \\"refSphere2\\"
{
    double3 xformOp:translate = (4, 5, 6)
    uniform token[] xformOpOrder = [\\"xformOp:translate\\"]

    def Sphere \\"world\\"
    {
        color3f[] primvars:displayColor = [(1, 0, 0)]
        double radius = 2
    }
}

"
`;

exports[`USD Tutorials Referencing Layers 2`] = `
"#usda 1.0

over \\"refSphere\\" (
    prepend references = @./HelloWorld.usda@
)
{
    uniform token[] xformOpOrder = []
}

over \\"refSphere2\\" (
    prepend references = @./HelloWorld.usda@
)
{
    over \\"world\\"
    {
        color3f[] primvars:displayColor = [(1, 0, 0)]
    }
}

"
`;

exports[`USD Tutorials Simple Shading in USD 1`] = `
"#usda 1.0
(
    doc = \\"\\"\\"Generated from Composed Stage of root layer //SimpleShading.usda
\\"\\"\\"
    upAxis = \\"Y\\"
)

def Xform \\"TexModel\\" (
    kind = \\"component\\"
)
{
    def Mesh \\"card\\"
    {
        float3[] extent = [(-430, -145, 0), (430, 145, 0)]
        int[] faceVertexCounts = [4]
        int[] faceVertexIndices = [0, 1, 2, 3]
        rel material:binding = </TexModel/boardMat>
        point3f[] points = [(-430, -145, 0), (430, -145, 0), (430, 145, 0), (-430, 145, 0)]
        texCoord2f[] primvars:st = [(0, 0), (1, 0), (1, 1), (0, 1)] (
            interpolation = \\"varying\\"
        )
    }

    def Material \\"boardMat\\"
    {
        token inputs:frame:stPrimvarName = \\"st\\"
        token outputs:surface.connect = </TexModel/boardMat/PBRShader.outputs:surface>

        def Shader \\"stReader\\"
        {
            uniform token info:id = \\"UsdPrimvarReader_float2\\"
            token inputs:varname.connect = </TexModel/boardMat.inputs:frame:stPrimvarName>
            float2 outputs:result
        }

        def Shader \\"texture\\"
        {
            uniform token info:id = \\"UsdUVTexture\\"
            asset inputs:file = @USDLogoHeader.png@
            float2 inputs:st.connect = </TexModel/boardMat/stReader.outputs:result>
            float3 outputs:rgb
        }

        def Shader \\"PBRShader\\"
        {
            uniform token info:id = \\"UsdPreviewSurface\\"
            color3f inputs:diffuseColor.connect = </TexModel/boardMat/texture.outputs:rgb>
            float inputs:metallic = 0
            float inputs:roughness = 0.4
            token outputs:surface
        }
    }
}

"
`;
